src/clj/puppetlabs/services/file_serving/file_metadata/file_metadata_core.clj

```diff
@@ -0,0 +1,200 @@
+      (get-base-metadata file-path posix-attrs)
+      (assoc-in [:data :relative_path] nil)
+      (assoc-in [:data :destination] nil)
+      (assoc-in [:data :type] "file")
+      ))
+
+(defn- get-directory-metadata
+  [file-path posix-attrs]
+  (-> (assoc-in base-metadata [:data :checksum] (checksum-of-dir posix-attrs))
+      (get-base-metadata file-path posix-attrs)
+      (assoc-in [:data :relative_path] ".")
+      (assoc-in [:data :destination] nil)
+      (assoc-in [:data :type] "directory")
+      ))
+
+(defn- get-link-metadata
+  [file-path posix-attrs links]
+  (-> (assoc-in base-metadata [:data :checksum] (checksum-of-file file-path))
+      (get-base-metadata file-path posix-attrs)
+      (assoc-in [:data :relative_path] nil)
+      (assoc-in [:data :destination] (.getCanonicalPath (.toFile (posix-utils/resolve-link file-path))))
+      (assoc-in [:data :type] "link")
+      ))
+
+(def Link schema/Str)
+(def SourcePermission schema/Str)
+
+(schema/defn get-metadata
+  [mountpath :- config/MountPath
+   links :- Link
+   source_permissions :- SourcePermission]
+  (let [[mount path] mountpath
+        file (clojure.java.io/file (:path mount) path)
+        file-path (.toPath file)
+        posix-attrs (posix-utils/unix-attributes file-path links)]
+
+    (assoc-in (cond
+                (.isDirectory file) (get-directory-metadata file-path posix-attrs)
+                (java.nio.file.Files/isSymbolicLink file-path) (get-link-metadata file-path posix-attrs links)
+                :else (get-file-metadata file-path posix-attrs)) [:data :links] links)))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Compojure app
+
+(defn try-to-parse
+  [body]
+  (try
+    (cheshire/parse-stream (io/reader body) true)
+    (catch Exception e
+      (log/debug e))))
+
+(def media-types
+  #{"application/json" "text/pson" "pson"})
+
+(defn content-type-valid?
+  [context]
+  (let [content-type (get-in context [:request :headers "content-type"])]
+    (or
+      (nil? content-type)
+      (media-types content-type))))
+
+(defn as-json-or-pson
+  "This is a stupid hack because of PSON.  We shouldn't have to do this, but
+  liberator does not know how to serialize a map as PSON (as it does with JSON),
+  so we have to tell it how."
+  [x context]
+  (let [context-with-media-type (if (string/blank? (get-in context
+                                                           [:representation
+                                                            :media-type]))
+                                  (assoc-in context
+                                            [:representation :media-type]
+                                            "text/pson")
+                                  context)]
+    (-> (cheshire/generate-string x)
+        (representation/as-response context-with-media-type)
+        (assoc :status 200)
+        (representation/ring-response))))
```

Comment:
These PSON helper methods should be moved out into their own namespace, or maybe into `liberator-utils`, so they're not just copied from the CA code.